File: react-wasm-sudoku/generate.py
import sys
import random

def neighbors(x, y):
    for other_x in range(9):
        for other_y in range(9):
            if x == other_x and y == other_y:
                continue

            if x == other_x:
                yield (other_x, other_y)
                continue

            if y == other_y:
                yield (other_x, other_y)
                continue

            if (other_x // 3) == (x // 3) and (other_y // 3) == (y // 3):
                yield (other_x, other_y)

NEIGHBORS = {}

for x in range(9):
    for y in range(9):
        NEIGHBORS[(x, y)] = list(neighbors(x, y))

class SudokuGrid:
    def __init__(self):
        self.cells = [ [None] * 9 for _ in range(9) ]

    def render(self):
        return '\n'.join(''.join(str(cell) if cell else ' ' for cell in row) for row in self.cells)

    def _generate_helper(self, cell_order):
        # find all cells that still need a value
        remainder = [ cell for cell in cell_order if self.cells[cell[0]][cell[1]] is None ]

        if not remainder:
            return True

        # find the valid values for each
        valid_values = {}
        for cell in remainder:
            cell_values = {1, 2, 3, 4, 5, 6, 7, 8, 9}
            for other_x, other_y in NEIGHBORS[cell]:
                if v := self.cells[other_x][other_y]:
                    if v in cell_values:
                        cell_values.remove(v)
            valid_values[cell] = cell_values

        # find the cell with the lowest valid value count
        next_cell = min(remainder, key=lambda cell: len(valid_values[cell]))

        # find the value that affects the highest number of neighbors
        candidates = []
        for value in valid_values[next_cell]:
            remove_count = 0
            for other_cell in NEIGHBORS[next_cell]:
                if other_cell in valid_values and value in valid_values[other_cell]:
                    remove_count += 1
            candidates.append((value, remove_count))

        candidates.sort(key=lambda c: c[1])
        candidates.reverse()
        candidates = [ c[0] for c in candidates ]

        # recurse, stopping if we found a solution
        for value in candidates:
            self.cells[next_cell[0]][next_cell[1]] = value

            if self._generate_helper(cell_order):
                return True

        self.cells[next_cell[0]][next_cell[1]] = None
        return False

    def generate(self, seed=None):
        rng = random.Random(seed)

        cells_to_fill = [ (x, y) for x in range(9) for y in range(9) ]
        valid_values = { cell:{1, 2, 3, 4, 5, 6, 7, 8, 9} for cell in cells_to_fill }
        rng.shuffle(cells_to_fill)

        #assert self._generate_helper(cells_to_fill, 0, valid_values), 'no solution found'
        assert self._generate_helper(cells_to_fill), 'no solution found'

if len(sys.argv) > 1:
    seed = int(sys.argv[1])
else:
    import time
    seed = int(time.time())

g = SudokuGrid()
g.generate(seed=seed)
print(g.render())
print(seed)

# 1670605882 finishes quickly
# 1670606006 doesn't finish
# 1670608339 finishes in 1s
# 1670608393 finishes in 3s

File: react-wasm-sudoku/generate.go
package main

import (
	"math/rand"
	"sort"
	"strconv"
	"strings"
)

type position struct {
	i int
	j int
}

var Neighbors map[position][]position

func neighbors(x, y int) []position {
	n := make([]position, 0, 20) // 20 happens to be the maximum number of neighbors you can have

	for otherX := 0; otherX < 9; otherX++ {
		for otherY := 0; otherY < 9; otherY++ {
			if x == otherX && y == otherY {
				continue
			}

			if x == otherX {
				n = append(n, position{otherX, otherY})
				continue
			}

			if y == otherY {
				n = append(n, position{otherX, otherY})
				continue
			}

			if (otherX/3) == (x/3) && (otherY/3) == (y/3) {
				n = append(n, position{otherX, otherY})
				continue
			}
		}
	}

	return n
}

func init() {
	Neighbors = make(map[position][]position)

	for x := 0; x < 9; x++ {
		for y := 0; y < 9; y++ {
			Neighbors[position{x, y}] = neighbors(x, y)
		}
	}
}

type sudokuGrid struct {
	cells [9][9]int
}

func (g *sudokuGrid) String() string {
	lines := make([]string, 9)

	for i := 0; i < 9; i++ {
		cols := make([]string, 9)
		for j := 0; j < 9; j++ {
			if g.cells[i][j] == 0 {
				cols[j] = " "
			} else {
				cols[j] = strconv.Itoa(g.cells[i][j])
			}
		}
		lines[i] = strings.Join(cols, "")
	}

	return strings.Join(lines, "\n")
}

func (g *sudokuGrid) generateHelper(cells []position) bool {
	remainder := make([]position, 0, 81)
	for _, cell := range cells {
		if g.cells[cell.i][cell.j] == 0 {
			remainder = append(remainder, cell)
		}
	}

	if len(remainder) == 0 {
		return true
	}

	validValues := make(map[position]map[int]struct{})
	for _, cell := range remainder {
		cellValues := map[int]struct{}{
			1: {},
			2: {},
			3: {},
			4: {},
			5: {},
			6: {},
			7: {},
			8: {},
			9: {},
		}

		for _, otherCell := range Neighbors[cell] {
			delete(cellValues, g.cells[otherCell.i][otherCell.j])
		}

		validValues[cell] = cellValues
	}

	var nextCell position
	lowestValidValues := 10
	for _, cell := range remainder {
		cellValues := validValues[cell]
		if len(cellValues) < lowestValidValues {
			nextCell = cell
			lowestValidValues = len(cellValues)
		}
	}

	candidates := make([]int, 0, len(validValues[nextCell]))
	for v := range validValues[nextCell] {
		candidates = append(candidates, v)
	}

	sort.Ints(candidates)

	for _, value := range candidates {
		g.cells[nextCell.i][nextCell.j] = value

		if g.generateHelper(cells) {
			return true
		}
	}

	g.cells[nextCell.i][nextCell.j] = 0

	return false
}

func (g *sudokuGrid) generate(seed int64, numBlanks int) {
	r := rand.New(rand.NewSource(seed))
	cells := make([]position, 0, 81)
	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			cells = append(cells, position{i, j})
		}
	}

	r.Shuffle(len(cells), func(i, j int) {
		cells[i], cells[j] = cells[j], cells[i]
	})

	// XXX assert it's true?
	g.generateHelper(cells)

	i := 0
	for numBlanks > 0 {
		cell := cells[i]
		oldValue := g.cells[cell.i][cell.j]
		g.cells[cell.i][cell.j] = 0

		i++

		// if there's more than one solution it's not a valid puzzle, so restore the old value
		// and move on to the next
		if g.numSolutions() > 1 {
			g.cells[cell.i][cell.j] = oldValue
			continue
		}

		numBlanks--
	}
}

func (g *sudokuGrid) numSolutions() int {
	return len(g.findAllSolutions())
}

func (g *sudokuGrid) findAllSolutionsHelper(accum *[]sudokuGrid) {
	// find cells that still need a value
	blankCells := make([]position, 0)
	for i, row := range g.cells {
		for j, value := range row {
			if value == 0 {
				blankCells = append(blankCells, position{i, j})
			}
		}
	}

	// if there are none, we've found a solution
	if len(blankCells) == 0 {
		solution := sudokuGrid{}
		for i, row := range g.cells {
			for j, value := range row {
				solution.cells[i][j] = value
			}
		}

		*accum = append(*accum, solution)
		return
	}

	// calculate valid values for each one
	validValues := make(map[position]map[int]struct{})
	for _, cell := range blankCells {
		cellValues := map[int]struct{}{
			1: {},
			2: {},
			3: {},
			4: {},
			5: {},
			6: {},
			7: {},
			8: {},
			9: {},
		}

		for _, otherCell := range Neighbors[cell] {
			delete(cellValues, g.cells[otherCell.i][otherCell.j])
		}

		validValues[cell] = cellValues
	}

	// find the cell that has the lowest number of valid values
	var nextCell position
	lowestValidValues := 10
	for _, cell := range blankCells {
		cellValues := validValues[cell]
		if len(cellValues) < lowestValidValues {
			nextCell = cell
			lowestValidValues = len(cellValues)
		}
	}

	// for each value, fill in the cell with it and recurse to move on to the next
	for value := range validValues[nextCell] {
		g.cells[nextCell.i][nextCell.j] = value

		g.findAllSolutionsHelper(accum)
	}

	// restore the blank value before returning control to caller to backtrack
	g.cells[nextCell.i][nextCell.j] = 0
}

func (g *sudokuGrid) findAllSolutions() []sudokuGrid {
	var solutions []sudokuGrid

	g.findAllSolutionsHelper(&solutions)

	return solutions
}
